File: Tetris-CLI/Tetris.py
# Tetris Game in Python - ASCII Version. By Yohan
# This is an ASCII Tetris game implemented in Python.
# It uses the terminal for rendering and keyboard input.
# The game features three boards (left, center, right) and allows switching between them.

import os
import random
import select
import shutil
import sys
import termios
import time
import tty

WIDTH, HEIGHT = 10, 24
FALL_SPEED = 0.5

BOARD_LEFT = [[" " for _ in range(WIDTH)] for _ in range(HEIGHT)]
BOARD_CENTER = [[" " for _ in range(WIDTH)] for _ in range(HEIGHT)]
BOARD_RIGHT = [[" " for _ in range(WIDTH)] for _ in range(HEIGHT)]
BOARDS = {"left": BOARD_LEFT, "center": BOARD_CENTER, "right": BOARD_RIGHT}

SHAPES = {
    "I": [[1, 1, 1, 1]],
    "O": [[1, 1], [1, 1]],
    "T": [[0, 1, 0], [1, 1, 1]],
    "S": [[0, 1, 1], [1, 1, 0]],
    "Z": [[1, 1, 0], [0, 1, 1]],
    "J": [[1, 0, 0], [1, 1, 1]],
    "L": [[0, 0, 1], [1, 1, 1]],
}

current_board = "center"
points = 0
fall_speed = FALL_SPEED


def rotate(shape):
    return [
        [shape[row][col] for row in range(len(shape) - 1, -1, -1)]
        for col in range(len(shape[0]))
    ]


def clear_screen():
    os.system("clear" if os.name == "posix" else "cls")


def draw_shape(shape, x, y, board):
    temp_board = [row.copy() for row in board]
    for i, row in enumerate(shape):
        for j, cell in enumerate(row):
            if cell:
                bx, by = x + j, y + i
                if 0 <= by < HEIGHT and 0 <= bx < WIDTH:
                    temp_board[by][bx] = "#"
    return temp_board


def render_board(board):
    return "\n".join(
        "|" + "".join("[]" if cell == "#" else "  " for cell in row) + "|"
        for row in board
    )


def display_game(shape=None, x=0, y=0, next_shape=None):
    global points
    temp = (
        draw_shape(shape, x, y, BOARDS[current_board])
        if shape
        else BOARDS[current_board]
    )

    left = render_board(temp if current_board == "left" else BOARD_LEFT)
    center = render_board(temp if current_board == "center" else BOARD_CENTER)
    right = render_board(temp if current_board == "right" else BOARD_RIGHT)

    term_width = shutil.get_terminal_size((80, 24))[0]
    offset = max((term_width - (WIDTH * 6 + 6)) // 2, 0)

    combined = "\n".join(
        " " * offset + l + "   " + c + "   " + r
        for l, c, r in zip(left.split("\n"), center.split("\n"), right.split("\n"))
    )

    clear_screen()
    print("\n" * 2)
    indicator = {
        "left": (WIDTH) * " " + "\\/",
        "center": (WIDTH * 3) * " " + "     \\/",
        "right": (WIDTH * 5) * " " + "          \\/",
    }
    print(" " * offset + indicator[current_board])

    print(" " * offset + ("+" + "--" * WIDTH + "+   ") * 3)
    print(combined)
    print(" " * offset + ("+" + "--" * WIDTH + "+   ") * 3)

    if next_shape:
        preview = [[" " for _ in range(4)] for _ in range(4)]
        for i, row in enumerate(next_shape):
            for j, cell in enumerate(row):
                if cell:
                    preview[i][j] = "#"
        preview_offset = offset + WIDTH * 3
        print(" " * preview_offset + f"Points: {points}")
        print(" " * preview_offset + "+" + "--" * 4 + "+")
        print(" " * preview_offset + "|Next Set|")
        print(" " * preview_offset + "+" + "--" * 4 + "+")
        for row in preview:
            print(
                " " * preview_offset
                + "|"
                + "".join("[]" if c == "#" else "  " for c in row)
                + "|"
            )
        print(" " * preview_offset + "+" + "--" * 4 + "+")

    print(
        "\n"
        + " " * (offset)
        + "Controls: [a] left  [d] right  [s] drop  [w/r] rotate  [q/e] switch  [x] quit"
    )


def can_move(shape, x, y, board):
    for i, row in enumerate(shape):
        for j, cell in enumerate(row):
            if cell:
                bx, by = x + j, y + i
                if not (0 <= bx < WIDTH and 0 <= by < HEIGHT):
                    return False
                if board[by][bx] != " ":
                    return False
    return True


def place_shape(shape, x, y, board):
    for i, row in enumerate(shape):
        for j, cell in enumerate(row):
            if cell:
                bx, by = x + j, y + i
                if 0 <= bx < WIDTH and 0 <= by < HEIGHT:
                    board[by][bx] = "#"


def clear_lines(board):
    global points, fall_speed
    new_board = [row for row in board if " " in row]
    lines_cleared = HEIGHT - len(new_board)
    while len(new_board) < HEIGHT:
        new_board.insert(0, [" "] * WIDTH)
    board[:] = new_board

    if lines_cleared:
        points += [0, 100, 300, 500, 800][min(lines_cleared, 4)]
        if points // 1000 > (points - lines_cleared * 100) // 1000:
            fall_speed = max(0.01, fall_speed * 0.9)


def random_board_switch():
    global current_board
    current_board = random.choice(["left", "center", "right"])


def maybe_add_obstacles(last_time, interval):
    if time.time() - last_time > interval:
        random_board_switch()
        return time.time()
    return last_time


def read_key(timeout=0.1):
    fd = sys.stdin.fileno()
    old = termios.tcgetattr(fd)
    try:
        tty.setcbreak(fd)
        r, _, _ = select.select([fd], [], [], timeout)
        return sys.stdin.read(1) if r else None
    finally:
        termios.tcsetattr(fd, termios.TCSADRAIN, old)


def switch_board(target, shape, x, y):
    global current_board
    if can_move(shape, x, y, BOARDS[target]):
        current_board = target
    else:
        print("Game Over!")
        sys.exit()


def main_game():
    global current_board, fall_speed
    current_board = "center"
    next_shape = random.choice(list(SHAPES.values()))

    last_obstacle_time = time.time()
    obstacle_interval = 10

    while True:
        shape = next_shape
        next_shape = random.choice(list(SHAPES.values()))
        x = WIDTH // 2 - len(shape[0]) // 2
        y = 0

        if not can_move(shape, x, y, BOARDS[current_board]):
            print("Game Over!")
            break

        fall_timer = time.time()
        while True:
            display_game(shape, x, y, next_shape)
            key = read_key(0.05)

            obstacle_interval = max(2, 10 - points // 1000)
            last_obstacle_time = maybe_add_obstacles(
                last_obstacle_time, obstacle_interval
            )

            if key == "a" and can_move(shape, x - 1, y, BOARDS[current_board]):
                x -= 1
            elif key == "d" and can_move(shape, x + 1, y, BOARDS[current_board]):
                x += 1
            elif key == "s":
                while can_move(shape, x, y + 1, BOARDS[current_board]):
                    y += 1
                place_shape(shape, x, y, BOARDS[current_board])
                clear_lines(BOARDS[current_board])
                break
            elif key == "w" or key == "r":
                r = rotate(shape)
                if can_move(r, x, y, BOARDS[current_board]):
                    shape = r
            elif key == "q":
                switch_board(
                    {"left": "right", "center": "left", "right": "center"}[
                        current_board
                    ],
                    shape,
                    x,
                    y,
                )
            elif key == "e":
                switch_board(
                    {"left": "center", "center": "right", "right": "left"}[
                        current_board
                    ],
                    shape,
                    x,
                    y,
                )
            elif key == "x":
                return

            if time.time() - fall_timer > fall_speed:
                if can_move(shape, x, y + 1, BOARDS[current_board]):
                    y += 1
                else:
                    place_shape(shape, x, y, BOARDS[current_board])
                    clear_lines(BOARDS[current_board])
                    break
                fall_timer = time.time()


if __name__ == "__main__":
    main_game()

File: Tetris-CLI/Tetris.go
package main

import (
	"fmt"
	"golang.org/x/term"
	"math/rand"
	"os"
	"os/exec"
	"runtime"
	"strings"
	"time"
)

const (
	BLOCK_CELL            = "[]"
	EMPTY_CELL            = "  "
	FPS               int = 24
	HEIGHT            int = 24
	HORIZONTAL_BORDER     = "-"
	VERTICAL_BORDER       = "|"
	WIDTH             int = 10
)

var board [HEIGHT][WIDTH]int
var boards map[string][HEIGHT][WIDTH]int = make(map[string][HEIGHT][WIDTH]int)
var clear map[string]func()
var currentBoard string = "center"
var currentPiece [][]int
var fallInterval float64 = 10
var gameTicker *time.Ticker
var gameWidth int = 0
var level int = 0
var linesForNextLevel int = 10
var nextPiece [][]int
var piecePosition [2]int
var rng *rand.Rand
var score float64
var terminalWidth int = 80
var totalLinesCleared int = 0
var userInputChan = make(chan rune, 1)

// BOARD OPERATIONS
func callClear() {
	if clearFunc, ok := clear[runtime.GOOS]; ok {
		clearFunc()
	} else {
		fmt.Println("Warning: Unsupported platform, cannot clear terminal screen.")
	}
}

func clearLines() {
	if currentBoard != "center" {
		return
	}
	tempScore := 0.0
	multiplier := 1.0
	linesCleared := 0

	for r := HEIGHT - 1; r >= 0; {
		isLineFull := true
		for c := range WIDTH {
			if board[r][c] == 0 {
				isLineFull = false
				break
			}
		}

		if isLineFull {
			linesCleared++
			for moveR := r; moveR > 0; moveR-- {
				copy(board[moveR][:], board[moveR-1][:])
			}
			board[0] = [WIDTH]int{}
		} else {
			r--
		}
	}

	if linesCleared > 0 {
		totalLinesCleared += linesCleared
		if totalLinesCleared >= linesForNextLevel {
			level++
			linesForNextLevel = (level + 1) * 10
			updateFallSpeedByLevel()
		}
		if linesCleared == 1 {
			tempScore = 100
		} else if linesCleared == 2 {
			tempScore = 300
		} else if linesCleared == 3 {
			tempScore = 500
		} else {
			tempScore = 850
		}
		boards[currentBoard] = board

		leftMultiplier := scoreMultipliers("left")
		rightMultiplier := scoreMultipliers("right")

		multiplier = (leftMultiplier * rightMultiplier)

		tempScore *= multiplier
		score += tempScore
		switchBoard("center")
	}
}

func drawBoard() {
	callClear()

	leftTempBoard := boards["left"]
	centerTempBoard := boards["center"]
	rightTempBoard := boards["right"]

	activeTempBoard := &leftTempBoard
	if currentBoard == "center" {
		activeTempBoard = &centerTempBoard
	} else if currentBoard == "right" {
		activeTempBoard = &rightTempBoard
	}

	if currentPiece != nil {
		for rowIdx, rowData := range currentPiece {
			for cellIdx, cell := range rowData {
				if cell == 1 {
					boardRow := piecePosition[0] + rowIdx
					boardCol := piecePosition[1] + cellIdx
					if boardRow >= 0 && boardRow < HEIGHT && boardCol >= 0 && boardCol < WIDTH {
						(*activeTempBoard)[boardRow][boardCol] = 1
					}
				}
			}
		}
	}

	borderLine := strings.Repeat(HORIZONTAL_BORDER, WIDTH*len(BLOCK_CELL))
	preview := strings.Split(drawNextPiece(), "\r\n")

	fmt.Println(centerString("Tetris CLI") + "\r")
	fmt.Println(centerString(fmt.Sprintf("Score: %.0f", score)) + "\r")
	fmt.Println(centerString(fmt.Sprint("Next Piece:")) + "\r")
	for _, line := range preview {
		fmt.Println(centerString(line) + "\r")
	}
	fmt.Println(centerString("+" + borderLine + "+ " + "+" + borderLine + "+ " + "+" + borderLine + "+\r"))

	for r := range HEIGHT {
		boardLine := ""
		boardLine += VERTICAL_BORDER
		for c := range WIDTH {
			if leftTempBoard[r][c] == 1 {
				boardLine += BLOCK_CELL
			} else {
				boardLine += EMPTY_CELL
			}
		}
		boardLine += VERTICAL_BORDER + " " + VERTICAL_BORDER

		for c := range WIDTH {
			if centerTempBoard[r][c] == 1 {
				boardLine += BLOCK_CELL
			} else {
				boardLine += EMPTY_CELL
			}
		}

		boardLine += VERTICAL_BORDER + " " + VERTICAL_BORDER

		for c := range WIDTH {
			if rightTempBoard[r][c] == 1 {
				boardLine += BLOCK_CELL
			} else {
				boardLine += EMPTY_CELL
			}
		}
		boardLine += VERTICAL_BORDER + "\r"
		fmt.Println(centerString(boardLine))
	}
	fmt.Println(centerString("+" + borderLine + "+ " + "+" + borderLine + "+ " + "+" + borderLine + "+\r"))
	fmt.Println(centerString("\nControls: WASD/HJKL = Movement | F = Hard Drop | Q/E = Switch Board | X = Exit"))
}

// PIECE OPERATIONS
func drawNextPiece() string {
	previewWidth := 4
	previewHeight := 2

	preview := make([][]string, previewHeight)
	for i := range preview {
		preview[i] = make([]string, previewWidth)
		for j := range preview[i] {
			preview[i][j] = EMPTY_CELL
		}
	}

	offsetX := (previewWidth - len(nextPiece[0])) / 2
	offsetY := (previewHeight - len(nextPiece)) / 2

	for y, row := range nextPiece {
		for x, cell := range row {
			if cell == 1 && y+offsetY < previewHeight && x+offsetX < previewWidth {
				preview[y+offsetY][x+offsetX] = BLOCK_CELL
			}
		}
	}

	var result strings.Builder
	result.WriteString("+" + strings.Repeat(HORIZONTAL_BORDER, previewWidth*len(BLOCK_CELL)) + "+\r\n")
	for _, row := range preview {
		result.WriteString(VERTICAL_BORDER)
		for _, cell := range row {
			result.WriteString(cell)
		}
		result.WriteString(VERTICAL_BORDER + "\r\n")
	}

	result.WriteString("+" + strings.Repeat(HORIZONTAL_BORDER, previewWidth*len(BLOCK_CELL)) + "+")
	return result.String()
}

func hardDrop() {
	for moveDown() {
	}
	lockPiece()
	spawnPiece()
}

// UI OPERATIONS
func centerString(s string) string {
	padding := max(0, (terminalWidth-len(s))/2)
	return strings.Repeat(" ", padding) + s
}

// SCORE OPERATIONS
func scoreMultipliers(boardName string) float64 {
	tempCurrentBoard := currentBoard
	switchBoard(boardName)

	linesCleared := 0
	for r := HEIGHT - 1; r >= 0; {
		isLineFull := true
		for c := range WIDTH {
			if board[r][c] == 0 {
				isLineFull = false
				break
			}
		}

		if isLineFull {
			linesCleared++
			for moveR := r; moveR > 0; moveR-- {
				copy(board[moveR][:], board[moveR-1][:])
			}
			board[0] = [WIDTH]int{}
		} else {
			r--
		}
	}

	boards[boardName] = board
	switchBoard(tempCurrentBoard)

	if linesCleared == 0 {
		return 1
	} else if linesCleared == 1 {
		return 1.5
	} else if linesCleared == 2 {
		return 2
	} else if linesCleared == 3 {
		return 2.5
	} else {
		return 3
	}
}

func init() {
	boards["left"] = [HEIGHT][WIDTH]int{}
	boards["center"] = [HEIGHT][WIDTH]int{}
	boards["right"] = [HEIGHT][WIDTH]int{}
	board = boards[currentBoard]
	gameWidth = 3*(WIDTH*len(BLOCK_CELL)+2) + 2
	rng = rand.New(rand.NewSource(time.Now().UnixNano()))
	score = 0
	clear = make(map[string]func())
	clear["linux"] = func() {
		cmd := exec.Command("clear")
		cmd.Stdout = os.Stdout
		cmd.Run()
	}
	clear["windows"] = func() {
		cmd := exec.Command("cmd", "/c", "cls")
		cmd.Stdout = os.Stdout
		cmd.Run()
	}
	updateWidth()
}

func isValidState(pieceToCheck [][]int, pos [2]int) bool {
	if pieceToCheck == nil {
		return false
	}
	for rowIdx, rowData := range pieceToCheck {
		for cellIdx, cell := range rowData {
			if cell == 1 {
				boardRow := pos[0] + rowIdx
				boardCol := pos[1] + cellIdx

				if boardRow < 0 || boardRow >= HEIGHT || boardCol < 0 || boardCol >= WIDTH {
					return false
				}
				if board[boardRow][boardCol] == 1 {
					return false
				}
			}
		}
	}
	return true
}

func listenForInput() {
	fd := int(os.Stdin.Fd())
	if !term.IsTerminal(fd) {
		fmt.Fprintln(os.Stderr, "Input is not a terminal. Keyboard input disabled.")
		close(userInputChan)
		return
	}

	oldState, err := term.MakeRaw(fd)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to set terminal to raw mode: %v\n", err)
		close(userInputChan)
		return
	}
	defer term.Restore(fd, oldState)

	buf := make([]byte, 1)
	for {
		n, err := os.Stdin.Read(buf)
		if err != nil || n == 0 {
			close(userInputChan)
			return
		}
		userInputChan <- rune(buf[0])
	}
}

func lockPiece() {
	if currentPiece == nil {
		return
	}
	for rowIdx, rowData := range currentPiece {
		for cellIdx, cell := range rowData {
			if cell == 1 {
				boardRow := piecePosition[0] + rowIdx
				boardCol := piecePosition[1] + cellIdx
				if boardRow >= 0 && boardRow < HEIGHT && boardCol >= 0 && boardCol < WIDTH {
					board[boardRow][boardCol] = 1
				}
			}
		}
	}
	clearLines()

	boards[currentBoard] = board
}

func moveDown() bool {
	nextPosition := [2]int{piecePosition[0] + 1, piecePosition[1]}
	if isValidState(currentPiece, nextPosition) {
		piecePosition = nextPosition
		return true
	}
	return false
}

func moveHorizontally(deltaX int) {
	nextPosition := [2]int{piecePosition[0], piecePosition[1] + deltaX}
	if isValidState(currentPiece, nextPosition) {
		piecePosition = nextPosition
	}
}

func rotate() {
	if currentPiece == nil || len(currentPiece) == 0 || len(currentPiece[0]) == 0 {
		return
	}

	originalRows := len(currentPiece)
	originalCols := len(currentPiece[0])

	rotatedPiece := make([][]int, originalCols)
	for i := range rotatedPiece {
		rotatedPiece[i] = make([]int, originalRows)
	}

	for r := range currentPiece {
		for c := range currentPiece[r] {
			rotatedPiece[c][originalRows-1-r] = currentPiece[r][c]
		}
	}

	if isValidState(rotatedPiece, piecePosition) {
		currentPiece = rotatedPiece
	}
}

func setNext() {
	shapes := map[string][][]int{
		"I": {{1, 1, 1, 1}},
		"O": {{1, 1}, {1, 1}},
		"T": {{0, 1, 0}, {1, 1, 1}},
		"S": {{0, 1, 1}, {1, 1, 0}},
		"Z": {{1, 1, 0}, {0, 1, 1}},
		"J": {{1, 0, 0}, {1, 1, 1}},
		"L": {{0, 0, 1}, {1, 1, 1}},
	}

	keys := make([]string, 0, len(shapes))
	for shape := range shapes {
		keys = append(keys, shape)
	}

	if len(keys) > 0 {
		randomKey := keys[rng.Intn(len(keys))]
		nextPiece = shapes[randomKey]
	} else {
		nextPiece = [][]int{{1}}
	}
}

func spawnPiece() {
	currentPiece = nextPiece
	if len(currentPiece) > 0 && len(currentPiece[0]) > 0 {
		piecePosition = [2]int{0, WIDTH/2 - len(currentPiece[0])/2}
	} else {
		piecePosition = [2]int{0, WIDTH / 2}
		currentPiece = [][]int{{1}}
	}

	if !isValidState(currentPiece, piecePosition) {
		callClear()
		drawBoard()
		os.Exit(0)
	}
	setNext()
}

func switchBoard(newBoard string) {
	boards[currentBoard] = board
	currentBoard = newBoard
	board = boards[currentBoard]
}

func updateFallSpeedByLevel() {
	baseFallSpeed := 10.0
	newSpeed := max(0.0, baseFallSpeed-(float64(level)*0.5))

	fallInterval = newSpeed

	if gameTicker != nil {
		gameTicker.Reset(time.Duration(float64(time.Second) * fallInterval / float64(FPS)))
	}
}

func updateWidth() {
	if width, _, err := term.GetSize(int(os.Stdout.Fd())); err == nil {
		terminalWidth = width
	}
}

func main() {
	setNext()
	spawnPiece()

	go listenForInput()

	drawBoard()

	gameTicker = time.NewTicker(time.Duration(float64(time.Second) * fallInterval / float64(FPS)))
	defer gameTicker.Stop()

	for {
		select {
		case <-gameTicker.C:
			if !moveDown() {
				lockPiece()
				spawnPiece()
			}
			drawBoard()

		case inputKey, ok := <-userInputChan:
			if !ok {
				fmt.Println("Input channel closed. Exiting.")
				return
			}

			actionTaken := true
			switch inputKey {
			case 'a', 'A', 'h', 'H':
				moveHorizontally(-1)
			case 'd', 'D', 'l', 'L':
				moveHorizontally(1)
			case 's', 'S', 'j', 'J':
				if !moveDown() {
					lockPiece()
					spawnPiece()
				}
			case 'f', 'F':
				hardDrop()
			case 'w', 'W', 'k', 'K':
				rotate()
			case 'q', 'Q':
				switch currentBoard {
				case "left":
					switchBoard("right")
				case "right":
					switchBoard("center")
				case "center":
					switchBoard("left")
				}
			case 'e', 'E':
				switch currentBoard {
				case "left":
					switchBoard("center")
				case "right":
					switchBoard("left")
				case "center":
					switchBoard("right")
				}
			case 'x', 'X':
				return
			default:
				actionTaken = false
			}

			if actionTaken {
				drawBoard()
			}
		}
	}
}
